### DISCUSSION TITLE
Should we remove size from the code-base?

### DISCUSSION THREAD

**Discussion by twiecki** [pymc-devs/pymc3#5790] [General]
It will already be excluded from the public API. Should we remove it?

Ref https://github.com/pymc-devs/pymc/issues/5754

**Comment by ricardoV94** [comment]
## Reasons to keep shape:
1. Users know shape
2. Dims agree with shape
3. Backwards compatibility with V3 (except in all the cases where shape was used inconsistently)

## Reasons to keep size:
1. Aesara and Numpy use size.
2. Internally all methods have to switch to size at some point (such as, but not limited to, `moment` and `RandomVariable.[call, make_node, infer_shape, rng_fn])`. It's a bit confusing to see shape in the signature of `__new__` and `dist` (or size explicitly forbidden), and then see size in all other function signatures related to the same variables.
3. We are not enforcing shape for support dimensionality and may never do
4. No extra functionality needed (shape is an extra)

# Reasons to not keep both (my strongest objection):
1. Less tests
2. Maintainers will have to keep remembering that users may pass one or the other whenever they create a variable and that our distributions should handle it properly. This is quite annoying for distribution factories like Mixtures and Timeseries.
3. Informal code of conduct to not mention size or shape **will not work**. It will show up in source-code, documentation, non-official code.
5. From 3. it follows that we have to document the omitted format.

# Reasons to keep both:
1. Initial one-time cost of removing any uses of shape/size in tests.
2. Backwards compatibility (will keep getting worse with future releases)
3. Devs are divided over which one to keep

**Reply by twiecki** [reply]
What option do you prefer?

**Reply by ricardoV94** [reply]
> What option do you prefer?

Size

**Reply by twiecki** [reply]
...and deprecate shape?

**Reply by ricardoV94** [reply]
Yeah I am more inclined to it the more I think about it

**Reply by reshamas** [reply]
Is there an informative error message that directs users to use `size` over `shape`?

**Reply by Armavica** [reply]
I was confused why the functions of `numpy.random` don't use `shape` like `np.zeros(shape=(3, 4))` but just realized that some distributions are over vectors, so for example `np.random.dirichlet([1, 1], size=(3, 4))` has a shape `(3, 4, 2)`. I think it is just a little unfortunate that the `.size` of an array is also its number of elements. For what it is worth, I think that anything is bound to be confusing, and that we should probably follow the principle of least surprise by following `numpy`.

**Reply by twiecki** [reply]
@Armavica Good points. However, there's also confusion caused by switching, but that is decreasing over time, while it otherwise stays.
